Mathlib/Probability/Kernel/Disintegration/Basic.lean

```diff
@@ -0,0 +1,518 @@
+  let e := embeddingReal Ω
+  let he := measurableEmbedding_embeddingReal Ω
+  let κ' := map κ (Prod.map (id : β → β) e) (measurable_id.prod_map he.measurable)
+  have hη' : fst κ' ⊗ₖ η = κ' := hη
+  have h_prod_embed : MeasurableEmbedding (Prod.map (id : β → β) e) :=
+    MeasurableEmbedding.id.prod_mk he
+  change fst κ ⊗ₖ borelMarkovFromReal Ω η = comapRight (fst κ' ⊗ₖ η) h_prod_embed
+  rw [comapRight_compProd_id_prod _ _ he]
+  have h_fst : fst κ' = fst κ := by
+    ext a u
+    unfold_let κ'
+    rw [fst_apply, map_apply, Measure.map_map measurable_fst h_prod_embed.measurable, fst_apply]
+    congr
+  rw [h_fst]
+  ext a t ht : 2
+  simp_rw [compProd_apply _ _ _ ht]
+  refine lintegral_congr_ae ?_
+  have h_ae : ∀ᵐ t ∂(fst κ a), (a, t) ∈ {p : α × β | η p (range e)ᶜ = 0} := by
+    rw [← h_fst]
+    have h_compProd : κ' a (univ ×ˢ range e)ᶜ = 0 := by
+      unfold_let κ'
+      rw [map_apply']
+      swap; · exact (MeasurableSet.univ.prod he.measurableSet_range).compl
+      suffices Prod.map id e ⁻¹' (univ ×ˢ range e)ᶜ = ∅ by rw [this]; simp
+      ext x
+      simp
+    rw [← hη', compProd_null] at h_compProd
+    swap; · exact (MeasurableSet.univ.prod he.measurableSet_range).compl
+    simp only [preimage_compl, mem_univ, mk_preimage_prod_right] at h_compProd
+    exact h_compProd
+  filter_upwards [h_ae] with a ha
+  rw [borelMarkovFromReal, comapRight_apply', comapRight_apply']
+  rotate_left
+  · exact measurable_prod_mk_left ht
+  · exact measurable_prod_mk_left ht
+  classical
+  rw [piecewise_apply, if_pos]
+  exact ha
+
+/-- For `κ' := map κ (Prod.map (id : β → β) e) (measurable_id.prod_map he.measurable)`, the
+hypothesis `hη` is `fst κ' ⊗ₖ η = κ'`. With that hypothesis,
+`fst κ ⊗ₖ borelMarkovFromReal κ η = κ`.-/
+lemma compProd_fst_borelMarkovFromReal (κ : kernel α (β × Ω)) [IsSFiniteKernel κ]
+    (η : kernel (α × β) ℝ) [IsSFiniteKernel η]
+    (hη : (fst (map κ (Prod.map (id : β → β) (embeddingReal Ω))
+        (measurable_id.prod_map (measurableEmbedding_embeddingReal Ω).measurable))) ⊗ₖ η
+      = map κ (Prod.map (id : β → β) (embeddingReal Ω))
+        (measurable_id.prod_map (measurableEmbedding_embeddingReal Ω).measurable)) :
+    fst κ ⊗ₖ borelMarkovFromReal Ω η = κ := by
+  let e := embeddingReal Ω
+  let he := measurableEmbedding_embeddingReal Ω
+  let κ' := map κ (Prod.map (id : β → β) e) (measurable_id.prod_map he.measurable)
+  have hη' : fst κ' ⊗ₖ η = κ' := hη
+  have h_prod_embed : MeasurableEmbedding (Prod.map (id : β → β) e) :=
+    MeasurableEmbedding.id.prod_mk he
+  have : κ = comapRight κ' h_prod_embed := by
+    ext c t : 2
+    unfold_let κ'
+    rw [comapRight_apply, map_apply, h_prod_embed.comap_map]
+  conv_rhs => rw [this, ← hη']
+  exact compProd_fst_borelMarkovFromReal_eq_comapRight_compProd κ η hη
+
+end BorelSnd
+
+section CountablyGenerated
+
+open ProbabilityTheory.kernel
+
+lemma isRatCondKernelCDFAux_density_Iic (κ : kernel α (γ × ℝ)) [IsFiniteKernel κ] :
```

Comment:
The next 70 lines are about kernels with values in a product where the second component is the real line, so they're interrupting a little bit the flow of the argument where you're moving things from the real line to a general standard Borel space. Would it make sense to move them to the top of the file (before the `BorelSnd` section)? Or to `CdfToKernel.lean`?